frame/merkle-mountain-range/rpc/src/lib.rs

```diff
@@ -185,6 +210,31 @@ where
 
 		Ok(true)
 	}
+
+	fn verify_proof_stateless(
+		&self,
+		mmr_root: MmrHash,
+		proof: LeavesProof<<Block as BlockT>::Hash>,
+	) -> RpcResult<bool> {
+		let api = self.client.runtime_api();
+
+		let leaves =
+			Decode::decode(&mut &proof.leaves.0[..]).map_err(runtime_error_into_rpc_error)?;
+
+		let decoded_proof =
+			Decode::decode(&mut &proof.proof.0[..]).map_err(runtime_error_into_rpc_error)?;
+
+		api.verify_proof_stateless(
+			&BlockId::hash(proof.block_hash),
+			mmr_root,
+			leaves,
+			decoded_proof,
+		)
+		.map_err(runtime_error_into_rpc_error)?
+		.map_err(mmr_error_into_rpc_error)?;
```

Comment:
This returns `bool` which is lost - this RPC will return `true` even for invalid proofs.

Reply:
If the proof is invalid due to incorrect MMR root, then [`verify_leaves_proof`](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/src/mmr/mmr.rs#L33-L37) returns `Ok(false)`, which [`verify_leaves`](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/src/lib.rs#L441-L445) maps to a `Result<(), primitives::Error>`, which is propagated up [here](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/rpc/src/lib.rs#L207-L208). So in this form, the rpc returns `true` only if `verify_leaves_proof` did, and the error returns early otherwise.
If the proof is invalid for other reasons than incorrect MMR root, then [`verify_leaves_proof`](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/src/mmr/mmr.rs#L33-L37) returns an error, which is propagated up regardless.

I agree with you though that this is not ideal in terms of type consistency, and this one's unfortunately a tradeoff. These are the other options I considered:

1.  have `verify_leaves` return `RpcResult<()>`. However, the json returns a `null` only then for the unwrapped success case, which in its raw form is misleading to an API consumer.
2.  use a wrapped `RpcResult<Result<(), CallError>>`, which smells antipattern to me
3.  map errors back to `false`, which sacrifices granularity for type consistency

Do you have another preference on how to handle this?